File: javaTool/src/main/java/nl/um/cds/triplifier/rdf/DataFactory.java
package nl.um.cds.triplifier.rdf;

import nl.um.cds.triplifier.rdf.ontology.DBO;
import org.apache.commons.lang.StringEscapeUtils;
import org.apache.log4j.Logger;
import org.eclipse.rdf4j.model.*;
import org.eclipse.rdf4j.model.vocabulary.RDF;
import org.eclipse.rdf4j.query.BindingSet;
import org.eclipse.rdf4j.query.TupleQueryResult;
import org.eclipse.rdf4j.query.Update;
import org.eclipse.rdf4j.repository.RepositoryResult;
import org.eclipse.rdf4j.rio.RDFHandler;
import org.eclipse.rdf4j.rio.ntriples.NTriplesWriter;

import java.io.FileOutputStream;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Properties;

public class DataFactory extends RdfFactory{
    private final OntologyFactory ontologyFactory;
    private final String baseIri;
    private static final Logger logger = Logger.getLogger(DataFactory.class);

    public DataFactory(OntologyFactory ontologyFactory, Properties props) {
        super(props);
        String hostname = super.getHostname();

        this.baseIri = "http://" + hostname + "/rdf/data/";
        this.ontologyFactory = ontologyFactory;
        this.initialize();
    }

    private void initialize(){
        this.initializeRdfStore();
        this.context = vf.createIRI("http://data.local/");
        this.conn.setNamespace("data", this.baseIri);
    }

    public void convertData() {
        try {
            Connection conn = this.connectDatabase(
                    this.getProperty("jdbc.driver"),
                    this.getProperty("jdbc.url"),
                    this.getProperty("jdbc.user"),
                    this.getProperty("jdbc.password"));

            TupleQueryResult tableList = this.ontologyFactory.getTablesFromOntology();
            while(tableList.hasNext()) {
                BindingSet tableObject = tableList.next();
                String tableClassUri = tableObject.getValue("tableClass").stringValue();
                String tableName = tableObject.getValue("tableName").stringValue();

                String catalog = null;
                if(tableObject.hasBinding("catalog")) {
                    catalog = tableObject.getValue("catalog").stringValue();
                }

                String schema = null;
                if(tableObject.hasBinding("schema")) {
                    schema = tableObject.getValue("schema").stringValue();
                }

                this.processTable(conn, tableClassUri, tableName, catalog, schema);
            }

        } catch (ClassNotFoundException | SQLException e) {
            e.printStackTrace();
        }

        //generateForeignKeyRelations();
    }

    private void generateForeignKeyRelations() {
        TupleQueryResult foreignKeys = ontologyFactory.getForeignKeys();
        while (foreignKeys.hasNext()) {
            BindingSet foreignKey = foreignKeys.next();
            String fkPredicate = foreignKey.getValue("fkPredicate").stringValue();
            String columnClassUri = foreignKey.getValue("columnClassUri").stringValue();
            String targetClassUri = foreignKey.getValue("targetClassUri").stringValue();

            String insertQuery = "PREFIX dbo: <http://um-cds/ontologies/databaseontology/>\n" +
                    "        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n" +
                    "        PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n" +
                    "\n" +
                    "        INSERT {\n" +
                    "            ?sources <"+fkPredicate+"> ?targets.\n" +
                    "        } WHERE {\n" +
                    "\n" +
                    "            ?sources rdf:type <"+columnClassUri+">;\n" +
                    "                     dbo:has_value ?columnValue.\n" +
                    "\n" +
                    "            ?targets rdf:type <"+targetClassUri+">;\n" +
                    "                     dbo:has_value ?columnValue.\n" +
                    "        }";
            logger.debug(insertQuery);
            Update update = this.conn.prepareUpdate(insertQuery);
            update.execute();
        }
    }

    private void processTable(Connection conn, String tableClassUri, String tableName, String catalog, String schema) {
        String query = "SELECT * FROM " + schema + "." + tableName;

        if(schema == null) {
            query = "SELECT * FROM " + tableName;
        }

        logger.info("Start processing table " + tableName);

        try {
            ResultSet sqlQueryResult = conn.prepareStatement(query).executeQuery();

            int resultRowId = 0;
            int numberRows = sqlQueryResult.getFetchSize();
            StatementCollector collector = new StatementCollector(this.context);

            while(sqlQueryResult.next()) {
                IRI tableRowIRI = this.getTableRowIRI(tableName, tableClassUri, resultRowId, sqlQueryResult);

                collector.addStatement(tableRowIRI, RDF.TYPE, vf.createIRI(tableClassUri));
                this.processColumns(sqlQueryResult, tableClassUri, tableRowIRI, collector);

                this.conn.add(collector.getStatements());
                collector.clearList();

                resultRowId++;
                logger.debug("Processed row " + resultRowId + " of " + numberRows);
            }
        } catch (SQLException e) {
            logger.warn("Could not execute query: " + query);
        }

    }

    public IRI getTableRowIRI(String tableName, String tableClassUri, int resultRowId, ResultSet queryResult) throws SQLException {
        String baseIriTable = this.baseIri + tableName + "/";
        IRI tableRowIRI = vf.createIRI(baseIriTable + resultRowId);

        TupleQueryResult pKeyList = this.ontologyFactory.getPrimaryKeysForTableFromOntology(tableClassUri);
        if(pKeyList.hasNext()) {
            String pKeyValue = "";
            boolean isFirst = true;
            while (pKeyList.hasNext()) {
                BindingSet pKey = pKeyList.next();
                if(!isFirst) {
                    pKeyValue += "_";
                } else {
                    isFirst = false;
                }

                String pKeyColumnName = pKey.getValue("columnName").stringValue();
                pKeyValue += queryResult.getString(pKeyColumnName);
            }
            tableRowIRI = vf.createIRI(baseIriTable + pKeyValue);
        }

        return tableRowIRI;
    }

    private void processColumns(ResultSet rowResults, String tableClassUri, IRI tableRowIRI, StatementCollector collector) throws SQLException {
        TupleQueryResult columnList = this.ontologyFactory.getColumnsForTableFromOntology(tableClassUri);
        while(columnList.hasNext()) {
            BindingSet columnResult = columnList.next();

            String columnClassUriString = columnResult.getValue("columnClassUri").stringValue();
            IRI columnClassUri = vf.createIRI(columnClassUriString);
            String columnName = columnResult.getValue("columnName").stringValue();

            IRI columnRowIRI = vf.createIRI(tableRowIRI.stringValue() + "/" + columnName.replaceAll(" ", "_"));
            String literalValue = rowResults.getString(columnName);

            collector.addStatement(columnRowIRI, RDF.TYPE, columnClassUri);
            collector.addStatement(tableRowIRI, DBO.HAS_COLUMN, columnRowIRI);
            // if there's no literal value for this column, then we can skip the creation of the column instance?
            if(literalValue != null) {
                IRI columnRowIRIValue = vf.createIRI(columnRowIRI.stringValue() + "/value");
                collector.addStatement(columnRowIRI, DBO.HAS_CELL, columnRowIRIValue);

                literalValue = StringEscapeUtils.escapeHtml(literalValue);
                collector.addStatement(columnRowIRIValue, RDF.TYPE, DBO.DATABASECELL);
                collector.addStatement(columnRowIRIValue, DBO.HAS_VALUE, vf.createLiteral(literalValue));
            }
        }
    }

    private Connection connectDatabase(String jdbcDriver, String jdbcUrl, String jdbcUser, String jdbcPass) throws SQLException, ClassNotFoundException {
        Class.forName(jdbcDriver);
        logger.debug("JDBC Driver loaded");

        Connection connection = DriverManager.getConnection(jdbcUrl, jdbcUser, jdbcPass);
        logger.debug("Connected to database");

        return connection;
    }

    public void exportData(String filePathName) throws IOException {
        RepositoryResult<Statement> result = this.conn.getStatements(null, null, null, true);
        FileOutputStream fos = new FileOutputStream(filePathName);

        RDFHandler nTriplesWriter = new NTriplesWriter(fos);
        this.conn.export(nTriplesWriter);

        fos.flush();
        fos.close();
    }
}

File: pythonTool/rdf/data_factory.py
from typing import Dict, List, Any
from rdflib import URIRef, Literal, RDF

from .rdf_factory import RdfFactory
from .statement_collector import StatementCollector
from .ontology.dbo import DBO, HAS_COLUMN, HAS_CELL, HAS_VALUE
from .ontology_factory import OntologyFactory
import sqlite3


class DataFactory(RdfFactory):
    def __init__(self, ontology_factory: OntologyFactory, props: Dict[str, str]):
        super().__init__(props)
        hostname = "localhost"
        self.base_iri = props.get("repo.dataUri", f"http://{hostname}/rdf/data/")
        self.ontology_factory = ontology_factory
        self.value_index: Dict[str, Dict[str, List[URIRef]]] = {}
        self.initialize()

    def initialize(self) -> None:
        self.initialize_rdf_store()
        self.context = URIRef(self.base_iri)
        self.graph.bind("data", self.base_iri)

    def convert_data(self) -> None:
        url = self.props.get("jdbc.url")
        if not url or not url.startswith("jdbc:sqlite:"):
            raise ValueError("Only sqlite connections are supported in this python port")
        path = url.replace("jdbc:sqlite:", "")
        conn = sqlite3.connect(path)

        for table_info in self.ontology_factory.tables.values():
            table_class_uri = table_info["class"]
            table_name = table_info["class"].split("/")[-1]
            self.process_table(conn, table_class_uri, table_name)

        self.generate_foreign_key_relations()

    def process_table(self, conn: sqlite3.Connection, table_class_uri: URIRef, table_name: str) -> None:
        cursor = conn.cursor()
        cursor.execute(f"SELECT * FROM {table_name}")
        columns = [d[0] for d in cursor.description]
        row_id = 0
        for row in cursor.fetchall():
            row_iri = self.get_table_row_iri(table_name, table_class_uri, row_id, columns, row)
            collector = StatementCollector(self.context)
            collector.add_statement(row_iri, RDF.type, table_class_uri)
            self.process_columns(row, columns, table_class_uri, row_iri, collector)
            self.add_statements(collector.get_statements())
            collector.clear()
            row_id += 1

    def get_table_row_iri(
        self,
        table_name: str,
        table_class_uri: URIRef,
        row_id: int,
        columns: List[str],
        row: Any,
    ) -> URIRef:
        base_iri_table = self.base_iri + table_name + "/"
        iri = URIRef(base_iri_table + str(row_id))

        pk_columns = self.ontology_factory.tables[table_name]["primary_keys"]
        if pk_columns:
            pkey_value = "_".join(str(row[columns.index(pk)]) for pk in pk_columns)
            iri = URIRef(base_iri_table + pkey_value)
        return iri

    def process_columns(
        self,
        row: Any,
        columns: List[str],
        table_class_uri: URIRef,
        row_iri: URIRef,
        collector: StatementCollector,
    ) -> None:
        for col_name, value in zip(columns, row):
            column_class_uri = self.ontology_factory.get_class_for_column(
                table_class_uri.split("/")[-1], col_name
            )
            column_row_iri = URIRef(f"{row_iri}/{col_name}")
            collector.add_statement(column_row_iri, RDF.type, column_class_uri)
            collector.add_statement(row_iri, HAS_COLUMN, column_row_iri)
            if value is not None:
                value_node = URIRef(f"{column_row_iri}/value")
                collector.add_statement(column_row_iri, HAS_CELL, value_node)
                collector.add_statement(value_node, RDF.type, DBO.Cell)
                collector.add_statement(value_node, HAS_VALUE, Literal(str(value)))
                idx = self.value_index.setdefault(str(column_class_uri), {})
                idx.setdefault(str(value), []).append(row_iri)

    def generate_foreign_key_relations(self) -> None:
        for fk in self.ontology_factory.foreign_keys:
            source_idx = self.value_index.get(str(fk["source_class"]), {})
            target_idx = self.value_index.get(str(fk["target_class"]), {})
            for value, subjects in source_idx.items():
                targets = target_idx.get(value, [])
                for s in subjects:
                    for t in targets:
                        self.add_statement(s, fk["predicate"], t)

    def export_data(self, file_path: str) -> None:
        self.graph.serialize(destination=file_path, format="nt")
